### PR TITLE
Add Stable Diffusion 3

### PR DESCRIPTION
**Pull request by DN6**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md:20]:
```suggestion

Original model weights can be found here: [https://huggingface.co/stabilityai/](https://huggingface.co/stabilityai/). 

```

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    negative_prompt="",
+    num_inference_steps=28,
+    height=1024,
+    width=1024,
+    guidance_scale=7.0,
+).images[0]
+
+image.save("sd3_hello_world-8bit-T5.png")
+```
+
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md:148]:
```suggestion

You can find the end-to-end script [here](https://gist.github.com/sayakpaul/82acb5976509851f2db1a83456e504f1). 

```

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+# Warm Up
+prompt = "a photo of a cat holding a sign that says hello world",
+for _ in range(3):
+    _ = pipe(prompt=prompt, generator=torch.manual_seed(1))
+
+# Run Inference
+image = pipe(prompt=prompt, generator=torch.manual_seed(1)).images[0]
+image.save("sd3_hello_world.png")
+```
+
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md:187]:
```suggestion

Check out the full script [here](https://gist.github.com/sayakpaul/508d89d7aad4f454900813da5d42ca97). 

```

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md (hunk 1 above)
Comment (DN6) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md:20]: Was this meant to be `[here](https://huggingface.co/stabilityai/)`?

File: src/diffusers/models/transformers/transformer_sd3.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+from ...models.attention import JointTransformerBlock
+from ...models.attention_processor import Attention, AttentionProcessor
+from ...models.modeling_utils import ModelMixin
+from ...models.normalization import AdaLayerNormContinuous
+from ...utils import is_torch_version
+from ..embeddings import CombinedTimestepTextProjEmbeddings, PatchEmbed
+from .transformer_2d import Transformer2DModelOutput
+
+
+class SD3Transformer2DModel(ModelMixin, ConfigMixin, PeftAdapterMixin):
```
[the remark this answers is not in the document]
Comment (DN6) [reply src/diffusers/models/transformers/transformer_sd3.py:32]: Yeah can do.

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md (hunk 1 above)
Comment (sayakpaul) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md:20]: Yeah feel free to shorten it. I wanted to unfurl it.

**Comment by kadirnar** [comment]
@DN6 

I tried this pull request. But it produces a black image. Can you check it?

https://huggingface.co/spaces/kadirnar/Stable-Diffusion-3

File: src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py [hunk 5]
```diff
@@ -0,0 +1,886 @@
...
+            2 ** (len(self.vae.config.block_out_channels) - 1) if hasattr(self, "vae") and self.vae is not None else 8
+        )
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+        self.tokenizer_max_length = (
+            self.tokenizer.model_max_length if hasattr(self, "tokenizer") and self.tokenizer is not None else 77
+        )
+        self.default_sample_size = (
+            self.transformer.config.sample_size
+            if hasattr(self, "transformer") and self.transformer is not None
+            else 64
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:201]:
```suggestion
            else 128
```

File: src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py [hunk 6]
```diff
@@ -0,0 +1,886 @@
...
+    def __call__(
+        self,
+        prompt: Union[str, List[str]] = None,
+        prompt_2: Optional[Union[str, List[str]]] = None,
+        prompt_3: Optional[Union[str, List[str]]] = None,
+        height: Optional[int] = None,
+        width: Optional[int] = None,
+        num_inference_steps: int = 50,
+        timesteps: List[int] = None,
+        guidance_scale: float = 5.0,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:629]: should we set it to be `7.0` (I think that's the default they want)

**Comment by vladmandic** [comment]
no support for `from_single_file`?
like i said many times, majority of users prefer to be able to download a model first - especially since stabilityai provided pre-packaged single-file versions that they actually promote.
also, using `from_pretrained` to automatically download a model is of limited use here since user needs to agree to terms&conditions first.

**Comment by sayakpaul** [comment]
It will come with single-file support. The PR isn't fully ready yet.

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 7]
```diff
@@ -237,21 +240,19 @@ def set_default_attn_processor(self):
 
     @apply_forward_hook
     def encode(
-        self, x: torch.Tensor, return_dict: bool = True
+        self, x: torch.FloatTensor, return_dict: bool = True
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:243]:
```suggestion
        self, x: torch.Tensor, return_dict: bool = True
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 8]
```diff
@@ -237,21 +240,19 @@ def set_default_attn_processor(self):
...
     ) -> Union[AutoencoderKLOutput, Tuple[DiagonalGaussianDistribution]]:
         """
         Encode a batch of images into latents.
 
         Args:
-            x (`torch.Tensor`): Input batch of images.
+            x (`torch.FloatTensor`): Input batch of images.
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:249]:
```suggestion
            x (`torch.Tensor`): Input batch of images.
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 9]
```diff
@@ -321,7 +330,7 @@ def blend_h(self, a: torch.Tensor, b: torch.Tensor, blend_extent: int) -> torch.
             b[:, :, :, x] = a[:, :, :, -blend_extent + x] * (1 - x / blend_extent) + b[:, :, :, x] * (x / blend_extent)
         return b
 
-    def tiled_encode(self, x: torch.Tensor, return_dict: bool = True) -> AutoencoderKLOutput:
+    def tiled_encode(self, x: torch.FloatTensor, return_dict: bool = True) -> AutoencoderKLOutput:
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:333]:
```suggestion
    def tiled_encode(self, x: torch.Tensor, return_dict: bool = True) -> AutoencoderKLOutput:
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 10]
```diff
@@ -331,15 +340,14 @@ def tiled_encode(self, x: torch.Tensor, return_dict: bool = True) -> Autoencoder
         output, but they should be much less noticeable.
 
         Args:
-            x (`torch.Tensor`): Input batch of images.
+            x (`torch.FloatTensor`): Input batch of images.
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:343]:
```suggestion
            x (`torch.Tensor`): Input batch of images.
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 11]
```diff
@@ -376,12 +384,12 @@ def tiled_encode(self, x: torch.Tensor, return_dict: bool = True) -> Autoencoder
 
         return AutoencoderKLOutput(latent_dist=posterior)
 
-    def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.Tensor]:
+    def tiled_decode(self, z: torch.FloatTensor, return_dict: bool = True) -> Union[DecoderOutput, torch.FloatTensor]:
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:387]:
```suggestion
    def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.FloatTensor]:
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 12]
```diff
@@ -376,12 +384,12 @@ def tiled_encode(self, x: torch.Tensor, return_dict: bool = True) -> Autoencoder
...
         r"""
         Decode a batch of images using a tiled decoder.
 
         Args:
-            z (`torch.Tensor`): Input batch of latent vectors.
+            z (`torch.FloatTensor`): Input batch of latent vectors.
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:392]:
```suggestion
            z (`torch.Tensor`): Input batch of latent vectors.
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 13]
```diff
@@ -426,14 +434,14 @@ def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[Decod
 
     def forward(
         self,
-        sample: torch.Tensor,
+        sample: torch.FloatTensor,
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:437]:
```suggestion
        sample: torch.Tensor,
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 14]
```diff
@@ -426,14 +434,14 @@ def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[Decod
...
         sample_posterior: bool = False,
         return_dict: bool = True,
         generator: Optional[torch.Generator] = None,
-    ) -> Union[DecoderOutput, torch.Tensor]:
+    ) -> Union[DecoderOutput, torch.FloatTensor]:
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:441]:
```suggestion
    ) -> Union[DecoderOutput, torch.Tensor]:
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 15]
```diff
@@ -426,14 +434,14 @@ def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[Decod
...
         r"""
         Args:
-            sample (`torch.Tensor`): Input sample.
+            sample (`torch.FloatTensor`): Input sample.
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:444]:
```suggestion
            sample (`torch.Tensor`): Input sample.
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py (hunk 13 above)
Comment (sayakpaul) [reply src/diffusers/models/autoencoders/autoencoder_kl.py:437]: I think we had decided to use `torch.FloatTensor` going forward

File: src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3_img2img.py [hunk 16]
```diff
@@ -0,0 +1,921 @@
...
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+
+        >>> from diffusers import AutoPipelineForImage2Image
+        >>> from diffusers.utils import load_image
+
+        >>> device = "cuda"
+        >>> model_id_or_path = "diffusers-internal-dev/private-model"
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3_img2img.py:60]: this should be updated later I guess

File: src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py [hunk 17]
```diff
@@ -0,0 +1,886 @@
...
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import StableDiffusion3Pipeline
+
+        >>> pipe = StableDiffusion3Pipeline.from_pretrained(
+        ...     "stabilityai/stable-diffusion-3-medium", torch_dtype=torch.float16
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3.py:58]: we need to update this

File: src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+    """
+    Output class for the scheduler's `step` function output.
+
+    Args:
+        prev_sample (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)` for images):
+            Computed sample `(x_{t-1})` of previous timestep. `prev_sample` should be used as next model input in the
+            denoising loop.
+        pred_original_sample (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)` for images):
+            The predicted denoised sample `(x_{0})` based on the model output from the current timestep.
+            `pred_original_sample` can be used to preview progress or for guidance.
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:41]:
```suggestion
```
my bad 😬

File: src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+
+    This model inherits from [`SchedulerMixin`] and [`ConfigMixin`]. Check the superclass documentation for the generic
+    methods the library implements for all schedulers such as loading and saving.
+
+    Args:
+        num_train_timesteps (`int`, defaults to 1000):
+            The number of diffusion steps to train the model.
+        timestep_spacing (`str`, defaults to `"linspace"`):
+            The way the timesteps should be scaled. Refer to Table 2 of the [Common Diffusion Noise Schedules and
+            Sample Steps are Flawed](https://huggingface.co/papers/2305.08891) for more information.
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:59]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+
+    def scale_noise(
+        self,
+        sample: torch.FloatTensor,
+        timestep: Union[float, torch.FloatTensor],
+        noise: Optional[torch.FloatTensor] = None,
+    ) -> torch.FloatTensor:
+        """
+        Ensures interchangeability with schedulers that need to scale the denoising model input depending on the
+        current timestep. Scales the denoising model input by `(sigma**2 + 1) ** 0.5` to match the Euler algorithm.
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:121]:
```suggestion
        Foward process in flow-matching
```

**Comment by kadirnar** [comment]
Will you write the inpaint pipeline code?

**Comment by GavChap** [comment]
requirements_sd3.txt is missing

File: examples/dreambooth/README_sd3.md [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+
+```bash
+git clone https://github.com/huggingface/diffusers
+cd diffusers
+pip install -e .
+```
+
+Then cd in the `examples/dreambooth` folder and run
+```bash
+pip install -r requirements_sd3.txt
```
Comment (apolinario) [inline examples/dreambooth/README_sd3.md:32]: This indeed seems to be missing

File: src/diffusers/loaders/single_file_utils.py [hunk 22]
```diff
@@ -96,6 +100,9 @@
         "pretrained_model_name_or_path": "stabilityai/stable-cascade-prior",
         "subfolder": "prior_lite",
     },
+    "sd3": {
+        "pretrained_model_name_or_path": "stabilityai/stable-diffusion-3-medium-diffusers-diffusers",
```
Comment (yiyixuxu) [inline src/diffusers/loaders/single_file_utils.py:104]:
```suggestion
        "pretrained_model_name_or_path": "stabilityai/stable-diffusion-3-medium-diffuser",
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 23]
```diff
@@ -376,7 +384,7 @@ def tiled_encode(self, x: torch.Tensor, return_dict: bool = True) -> Autoencoder
 
         return AutoencoderKLOutput(latent_dist=posterior)
 
-    def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.Tensor]:
+    def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.FloatTensor]:
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:387]:
```suggestion
    def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.Tensor]:
```

File: src/diffusers/models/autoencoders/autoencoder_kl.py [hunk 24]
```diff
@@ -262,19 +263,25 @@ def encode(
...
+
         posterior = DiagonalGaussianDistribution(moments)
 
         if not return_dict:
             return (posterior,)
 
         return AutoencoderKLOutput(latent_dist=posterior)
 
-    def _decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.Tensor]:
+    def _decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.FloatTensor]:
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl.py:278]:
```suggestion
    def _decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.Tensor]:
```

**Review by apolinario — APPROVED** [review]
wanted to make a comment and started a review by accident

File: docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md (hunk 2 above)
Comment (vladmandic) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md:156]: there's a typo: `"pretrained_model_name_or_path": "stabilityai/stable-diffusion-3-medium-diffusers-diffusers",`

-diffusers-diffusers

Comment (sayakpaul) [reply docs/source/en/api/pipelines/stable_diffusion/stable_diffusion_3.md:156]: We have updated it.

**Comment by GavChap** [comment]
I get this error as well when attempting to run the training (after figuring out the requirements)

```
An error occurred while trying to fetch stabilityai/stable-diffusion-3-medium-diffusers: stabilityai/stable-diffusion-3-medium-diffusers does not appear to have a file named diffusion_pytorch_model.safetensors.
Defaulting to unsafe serialization. Pass `allow_pickle=False` to raise an error instead.
Traceback (most recent call last):
  File "/home/me/diffusers/venv/lib/python3.10/site-packages/huggingface_hub/utils/_errors.py", line 304, in hf_raise_for_status
    response.raise_for_status()
  File "/home/me/diffusers/venv/lib/python3.10/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://huggingface.co/stabilityai/stable-diffusion-3-medium-diffusers/resolve/48ef03a6b806e781a368fe71b3855f515e789bb0/transformer/diffusion_pytorch_model.bin
```

**Comment by sayakpaul** [comment]
@GavChap I would appreciate if you don't post it in between an ongoing PR as the issue is not descriptive enough. So, I would suggest you wait for a little while and try it out. If you face any issue, open a separate thread.

**Pull request merged**